frontend/src/lib-components/atoms/buttons/AsyncButton.tsx

```diff
@@ -14,109 +20,157 @@ import { faCheck, faTimes } from 'lib-icons'
 
 import { StyledButton } from './Button'
 
-export type AsyncClickCallback = (
-  cancel: () => Promise<void>
-) => Promise<void | Result<unknown>>
+const onSuccessTimeout = isAutomatedTest ? 10 : 500
+const clearStateTimeout = isAutomatedTest ? 25 : 2000
 
-type Props = {
+type ButtonState<T> =
+  | { state: 'idle' | 'in-progress' | 'failure' }
+  | { state: 'success'; value: T }
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+const idle: ButtonState<any> = { state: 'idle' }
+const inProgress: ButtonState<any> = { state: 'in-progress' }
+const failure: ButtonState<any> = { state: 'failure' }
+
+/* eslint-enable @typescript-eslint/no-explicit-any */
+
+export interface Props<T> {
   text: string
   textInProgress?: string
   textDone?: string
-  onClick: AsyncClickCallback
-  onSuccess: () => void
-  onFailure?: (result?: Failure<unknown> | undefined) => void
+  onClick: () => Promise<Result<T>> | void
+  onSuccess: (value: T) => void
```

Comment:
Add `onSuccessImmediate` that runs right away instead after `onSuccessTimeout`

Reply:
This would have benefited only 1 or 2 of the existing AsyncButton usages, so I decided to not add it now. Let's add it later if it's needed.